### Model Best Depth
3
### Model Accuracy 
70.1%

### Justification
The model's accuracy reduces with an increase in the number of trees. An high  number of trees can result in model overfitting. Tree numbers below 4 result in lower scores due to the models inability to make good generalizations [underfitting].

### Best scores
#### depth => 4 = accuracy 70.2% =====> Best Score
#### depth => 3 = accuracy 69.8% =====> Second Best Score
#### depth => 5 = accuracy 69.6% =======> Third Best Score

In [1]:
import pandas as pd 
import numpy as np 

from sklearn.model_selection import cross_val_score

from sklearn.model_selection import RepeatedStratifiedKFold

from sklearn.tree import DecisionTreeClassifier

In [2]:
df = pd.read_csv('../datasets/dm_train_t.csv')


In [3]:
df.head()

id     n24_t   rev24_t  elpsdm_t_t_1  revlast_t_t_1  ordfreq  ordcat  \
0   1  0.693147  3.401197           0.0       4.472136        1       6   
1   2  3.218876  5.332719           0.0       4.472136        1       6   
2   3  1.609438  3.951244           0.0       3.872983        1       6   
3   4  2.397895  4.653960           0.0       3.872983        1       6   
4   5  0.693147  3.465736           0.0       4.123106        1       6   

   response  
0         1  
1         0  
2         0  
3         0  
4         0

In [4]:
df.isnull().any()

id               False
n24_t            False
rev24_t          False
elpsdm_t_t_1     False
revlast_t_t_1    False
ordfreq          False
ordcat           False
response         False
dtype: bool

In [5]:
df.columns 

Index(['id', 'n24_t', 'rev24_t', 'elpsdm_t_t_1', 'revlast_t_t_1', 'ordfreq',
       'ordcat', 'response'],
      dtype='object')

In [6]:
cols = ['n24_t', 'rev24_t', 'elpsdm_t_t_1', 'revlast_t_t_1', 'ordfreq','ordcat']

x = df[cols]
y = df['response']

In [19]:
model = DecisionTreeClassifier(criterion='entropy', max_depth=5)

In [20]:
cv = RepeatedStratifiedKFold(n_splits=10, n_repeats=3, random_state=1)


In [21]:
#evaluate the model
m_scores = cross_val_score(model, x, y, scoring='accuracy', cv=cv, n_jobs=-1)


In [22]:
#summarize the results
print("Accuracy: %.3f (%.3f)" % (np.mean(m_scores), np.std(m_scores)))

Accuracy: 0.699 (0.018)



depth => 3 = accuracy 70.1% =====> Best Score
depth => 5 = accuracy 69.9% =======> Second Best Score
depth => 4 = accuracy 69.8% =====> Third Best Score
depth => 6 = accuracy 69.1%
depth => 10 = accuracy 65.5%
depth => 15 = accuracy 61.6%
depth => 20 = accuracy 62.4%
depth => 23 = accuracy 60.2%



